### DISCUSSION TITLE
Proposal: allow generic constraints to be part of the signature for overloading resolution

### DISCUSSION THREAD

**Discussion by Ultrahead** [dotnet/csharplang#2013] [Language Ideas]
Currently, it is not allowed to declare the following methods on the same class:

```csharp
public static class MyTestClass1
{
        public static void MyTestMethod1<T>(this T t)
            where T : class
        { }

        public static void MyTestMethod1<T>(this T t)
            where T : struct
        { }
}
```

Although you are clearing stating that one method should receive reference types and the other one value types, you get the following compiling error: `Error CS0111 Type 'MyTestClass1' already defines a member called 'MyTestMethod1' with the same parameter types`.

As stated in #1628, a workaround is simple with C# 7.3: [to split those methods in different classes](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLRIMaOQSwG4SoAOsMECAdsgDQwg4A2APgAIAMABCwIwDcAsACghPAGxcATBwAa3IQG8hHZVwDMXbuIAqHAKIAPMlRwB7CgB4tAPgAUMABY5kHHTACUHAO73yEFxxAOAEkAZRMAWwggijIEMCgMCEsATyIIAHkKCCslFUVBFUKHBBNPDiyygDkTGCDwogYISJiIABMDRKIYUwobNwEClQBfIRHhQTFJGQkFXOUWdUmdAyNcM0tbByd/dy8fJH9A0IiolriEpK1UiC1PExzB5XzClWLS8ogqmrqGpogW9r6TrdMx9AaFMZjIQ4M7xRLBMKRaKxOGXKwceQcMYsKRXNKZPyY7G4663EwYrEiKSItEBBEnZHkVEbClQiaLKQ8ADss0eag04hYABYNAAOPpzDGSwp4KAIDj6bgcAC8HzKNJS+Kytn60pUsvlAEcVQruAA6Fb/Na9XV8mVyhVSVUVDgavE3O468EvZQGrwm/QSC2GK09MGSyFAA).

Now, what would happen if instead you declare the following methods ...?

```csharp
public static class MyTestClass2
{
        public static void MyTestMethod2<T>(this T t)
            where T : struct
        { }

        public static void MyTestMethod2<T>(this T t)
            where T : struct, Enum
        { }
}
```

... and then you attempt to split those methods in different classes? You will get the following error: `The call is ambiguous between the following methods or properties ...`. The solution: to enclose each class in its own namespace, which then leads me to the following issue: what would happen if somewhere in your code you reference both namespaces at top-most level? You will be getting again the same error due to ambiguity. Ditto, for the case when you find two generic methods with the same name, signature and constraints in different classes/namespaces. Of course, that this situation can be solved when you declare all classes and methods your-self, but when you consume many APIs from third-parties, workarounds can be found but at the expense of simplicity.

So my proposal here is two-folded:

1. To include generic constraints as part of the signature to (a) allow defining generic overloads on the same class/struct, and (b) improve conflict resolution, so that the most specific method is chosen by the compiler by default (in case of the last example above, when working with an enum, the second `MyTestMethod2` will be picked by the compiler because the constraint requests an struct which is also and Enum). And, 

2. To allow a way to easily select which method to use when the compiler cannot automatically resolve the conflict. For example, with an approach like the following ...:

```csharp
myField.{myNameSpace}.MyTestMethod(...);

myField.{myNameSpace.myClass}.MyTestMethod(...);

myField.{myNameSpace.myClass.myNestedClass}.MyTestMethod(...);

```
... and so on so forth.

**Comment by HaloFour** [comment]
@Ultrahead 

The CLR doesn't consider generic type constraints to be a part of the method signature thus you can't have two overloads that only differ by those constraints.  There's nothing much that C# can do here without CLR changes.

As for this explicit overload resolution, if the overloads are that convoluted I'd suggest using method names to disambiguate them.  That would already be substantially less verbose than what you've proposed here.

**Comment by Ultrahead** [comment]
> The CLR doesn't consider generic type constraints to be a part of the method signature thus you can't have two overloads that only differ by those constraints. There's nothing much that C# can do here without CLR changes.

Ok, then I hope that those in charge of the CLR consider including them as part of method signature.

> As for this explicit overload resolution, if the overloads are that convoluted I'd suggest using method names to disambiguate them. That would already be substantially less verbose than what you've proposed here.

Not necessarily since you could disambiguate them with something as simple as ...:

```csharp
myField.{className}.MyTestMethod(...);
```
... and yet still feeling that you are using an (extension) method.

**Comment by ufcpp** [comment]
Should this be listed in https://github.com/dotnet/csharplang/issues/317 ?

**Comment by HaloFour** [comment]
> Ok, then I hope that those in charge of the CLR consider including them as part of method signature.

I'd wager that the benefit of such overloading wouldn't benefit the cost of modifying the CLR.  This limitation has existed for the last 13 years and is at worst a minor impediment.  It's not like all methods need to have the same name.

I also think that it would be a breaking change to the IL language where all method calls are required to express the full signature of the target method.  I don't know that anyone cares to maintain backward/forward compatibility in the IL language itself, but I don't think that the CLR could be amended without additional work on the IL assemblers themselves.

**Comment by Ultrahead** [comment]
#317 What language proposals would benefit from CLR changes? Added there for consideration.

**Comment by yuridiniz** [comment]
This solution would help me a lot

I will give the most current example that I stumbled upon it

```c#
public abstract class BasicBusiness<T, TKey> where T : class, new()
{
    public async virtual Task<T> Get(TKey key) { /*...*/ }
    public async virtual Task<T> Remove(TKey key) { /*...*/ }
    /* etc */
}

public abstract class BasicBusiness<T> : BasicBusiness<T, int>  where T : EntityIntKey, new()
{
}

public abstract class BasicBusiness<T> : BasicBusiness<T, short>  where T : EntityShortKey, new()
{
}

public abstract class BasicBusiness<T> : BasicBusiness<T, Guid>  where T : EntityGuidKey, new()
{
}

```

This would have avoided me major refactoring efforts

**Comment by Denis535** [comment]
You could add to signature only information about reference or value type.
It would be enough for 99.9% cases.

Of course, this would not bring much benefit, but it would be useful sometimes. And for `nullable reference types` it would be really useful.
But it looks like you will never do it. This issue doesn't even have any label.

**Comment by Ultrahead** [comment]
I'd like to know whether this proposal has been considered in #317

**Comment by Ultrahead** [comment]
Was it listed in #317 ?

**Comment by sebas77** [comment]
I also wondered if it could be technically possible to add the generic parameters as part of a method signature. The reason is that regardless of how stupid the following example is, I find the following compiler behaviour inconsistent if I didn't know that the generic parameter doesn't make a signature unique. The first case compiles fine, while the second case generates an error as the method can be ambiguous, however, technically they both mean the same thing. Generic methods can and should be more powerful than they are now for generic type specialization and this feature could help it.

![image](https://user-images.githubusercontent.com/945379/139578358-7fe90f63-614f-4b45-a2a9-a6edc6beed41.png)
![image](https://user-images.githubusercontent.com/945379/139578362-62655978-2e36-47da-a14e-c17b49534d1e.png)

Not completely related, but interesting to highlight, the following won't be instead considered ambiguous

```
public static void Complete<T>(this T enumerator) where T: class, IEnumerator {    }
public static void Complete<T>(this T task, int randomOptionalParam = 0) where T : class, IEnumerator<TaskContract>    { }
```

because there is a c# rule that says that methods with optional parameters are considered "better"

**Reply by bernd5** [reply]
The constraint just says that the type T is convertible to `IEnumerator` / `IEnumerator<TaskContract>`. If you have a class for which both is valid it would be ambiguous.
And because `IEnumerator<TaskContract>` is based on `IEnumerator` the situation is always ambigous.
In the first case this is not the case because you force an interface type...

To get it more clear - just imagine how the overload set would look like when you replace T with a concrete type...

**Reply by PathogenDavid** [reply]
> To get it more clear - just imagine how the overload set would look like when you replace T with a concrete type...

This part isn't true. [The compiler will pick `IEnumerable<T>` when possible because it's considered a better overload](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD4AEBMBGAWACgsAGAAizwDoBhAewBtGBTAYwSnpgGcBuYmUp4ALAKKC8ANko5hAdmIBvYuTXCZWEeQCy0GAAoAlKvUqi6y+QCCyZBACeAGSg8E5JgBMXb8gF5yGBYAd2NxK3UfBAAeWAQAPnIAcxYg5Cg2KP9AkLDTCIAVFjcDLyijPnIAeiryAHUoZnI2CCaqAGYPADcWZEZ6CE98qyKSlLSM8sqa+sbGZtb5jtiYBO7e/sHhy1GEAwMOo3HeydcECuragFEADwAHRgyodxamhHphTvoevoGhi3UAF9hsMqJptLsDnhOqlUGBehBgKwjP5ElQAJwGABEADluABaY7pNjrX6DLEVUHSSgQ4p7ZZxRKw+H2JEsFF+NF4TFYgDiqROJO+Gz+FPEwKIQA). (Someone more familiar with [the overload resolution rules](https://github.com/dotnet/csharplang/blob/26ccd64e79722d274ec91ba9e3e4bbab82801184/spec/expressions.md#overload-resolution) can probably find the specific line of the spec, but I believe the logic here is `IEnumerator<T>` is considered more specific than `IEnumerator` since it extends it.)

You are right though that [it is ambiguous if the interfaces involved are disparate](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD4AEBMBGAWAChis8A2AAl2rwHZiBvYy12qrAFkoFloYAFAEoWbZkTaTKAIQD2CABaVg8pQF5KMAKYB3YQG5RUygBUtAZwQCVioYYlsAvkcouyHbmcsCAkgEFKCCFKNQA+WgBOAQAiANkANy1kABtZCAATaLs3CmpPCysfaWVgsMiY4oSk1Iys+2cSIlgEJIAzCABjLUp/SkYG5rbO7qK+hpo5RUoQHr84HuL+4iA==).

**Reply by sebas77** [reply]
@bernd5 I am not sure you are right, because `IEnumerator<TaskContract>` is an IEnumerator itself, regardless of its implementation, so they would be ambiguous too. If I understood correctly, because the parameters are part of the method signature, c# is able to pick up the best overload at compile-time, but for some reason, this cannot be done with generics (probably because they are not part of the signature)

**Reply by YairHalberstadt** [reply]
@PathogenDavid overload resolution spec currently doesn't work that way for constraints:

```csharp
using System;
using System.Collections;
using System.Collections.Generic;

 "".M();

public static class Extensions1
{
    public static void M<T>(this T enumerator) where T : IEnumerable
    {
    }
}

public static class Extensions2
{
    public static void M<T>(this T enumerator) where T : IEnumerable<char>
    {
    }
}
```

> error CS0121: The call is ambiguous between the following methods or properties: 'Extensions1.M<T>(T)' and 'Extensions2.M<T>(T)'

**Reply by PathogenDavid** [reply]
@YairHalberstadt You're right, I was more intending to respond to the last sentence about it not working in the non-generic case. I edited my comment to clarify that.

**Reply by timcassell** [reply]
We may need a new syntax to resolve ambiguities. Current ambiguities can be resolved by casting, but that wouldn't be sufficient to solve generic constraint ambiguities (and even if it would be sufficient, it would cause boxing of structs which defeats the purpose of generics).
